### DISCUSSION TITLE
Feature Request: Documentation of duckdb version vs storage format version

### DISCUSSION THREAD

**Discussion by igitur** [duckdb/duckdb#5730] [General]
Hi all,

Like many people, I encountered this error again:

> Trying to read a database file with version number 39, but we can only read version 38. 

I totally understand that the file storage format isn't stable yet. 

In this case, I'm happy to downgrade my version of duckdb to a version that support file format version 38, but it was quite difficult to determine what the highest duckdb version would be for version 38. I couldn't find any documentation. I browsed through the Releases page and search the summary of PR, but nothing was said there.

In the end, I had to browse the source code to find where the storage format version is stored ( https://github.com/duckdb/duckdb/blob/master/src/storage/storage_info.cpp ), find the relevant commit that changed it to 38 ( de5c683d8ea735efc1bf0d8c1179caf11dae4ab3 ), find the PR in which this was changed ( #4187 ) and then find the release that contains this PR ( https://github.com/duckdb/duckdb/releases/tag/v0.5.0 ). 

I'd like to request a documentation page, either on duckdb.org or a wiki page here that maps the duckdb versions to storage format versions, until the storage format is stable.

I hope that makes sense. 

Thanks for a great product

**Comment by Mytherin** [comment]
Thanks for the suggestion! Great idea. I have been meaning to do this for a while. We have documented the storage versions [here](https://duckdb.org/internals/storage), and I have also created a PR so that DuckDB tells you the version with which a specific file was created (#5758).

**Reply by danz68** [reply]
A few ages ago, when MonetDBlite was a thing, I had this "crazy" idea that once a data analysis project developemnt ends, I should not be bothered by any new versions features, fixes, db upgrades, a cycle which we are all aware that never ends. What did I do back then? Because the datasets I was dealing with were really large, I just incorporated the dll in the database, and first thing I was doing it before connecting to my db was to extract the dll, place it in a convenient path and use it. I was aware of the (limited to a couple of releases) incremental upgrade MonetDB was providing along its development cycle, but that was not good enough. 

To be honest, I will keep using this approach for a while, unless ... and this is something you may or you may not consider as an idea. 

The "storage" code is not that large. I think that after a certain release you can provide a perpetual storage reader/upgrader library, providing an optional functionality other apps are incorporating in the main app for keeping the compatibility with old versions. An automatic check_and_upgrade_if necessary (in case the conversion library is availble in the duckdb binaries path) would save a lot of troubles for many developers. In the meantime, saving the library with the project is the least we cand do. For this, a reduced storage library able to only save as parquet, for instance, would help reducing database/project overhead of the above mechanism.

**Reply by igitur** [reply]
Thanks @Mytherin . This helps a lot.

**Comment by igitur** [comment]
Hi, should https://duckdb.org/internals/storage be updated to mention v0.7.0 explicitly? Also, would it be possible to link to https://duckdb.org/internals/storage in the DuckDB documentation menu? I keep having to come back to this discussion to find the URL. I'm too old to memorise it :-)

**Reply by igitur** [reply]
And mentioning the file format version explicitly in the blog release [post](https://duckdb.org/2023/02/13/announcing-duckdb-070.html) would also be beneficial.
